vnc/vnc.c

```diff
@@ -1870,8 +1897,37 @@ lib_mod_connect(struct vnc *v)
 
             if (i != 0)
             {
-                v->server_msg(v, "VNC password failed", 0);
-                error = 2;
+                if (version >= 8)
+                {
+                    error = trans_force_read_s(v->trans, s, 4);
+                    if (error == 0)
+                    {
+                        in_uint32_be(s, i);
+                        if (i <= 1024)
+                        {
+                            error = trans_force_read_s(v->trans, s, i);
+                        }
+                        else
+                        {
+                            LOG(LOG_LEVEL_ERROR, "VNC security failure reason exceeds size limit (%d > 1024 bytes)", i);
+                            error = 1;
+                        }
+                    }
+                    else
+                    {
+                        i = 0;
+                    }
+                    if (error == 0)
+                    {
+                        in_str(s, err_reason, i);
+                        v->server_msg(v, err_reason, 0);
+                    }
```

Comment:
If we get to this point, the server has disconnected, and yet error is still 0 (it should probably be 2).

Also, `i` can be negative again.

Suggest something like:-

```c
            if (i != 0)
            {
                if (version < 8)
                {
                    g_snprintf(text, sizeof(text), "VNC password failed");
                }
                else
                {
                    g_snprintf(err_reason, sizeof(err_reason),
                               "<reason unavailable>");
                    if (trans_force_read_s(v->trans, s, 4) == 0)
                    {
                        in_uint32_be(s, i);
                        if (i > 0 && i < sizeof(err_reason) &&
                                trans_force_read_s(v->trans, s, i) == 0)
                        {
                            in_str(s, i, err_reason, sizeof(err_reason));
                        }
                    }
                    g_snprintf(text, sizeof(text), "VNC password failed : %s",
                               err_reason);
                }

                v->server_msg(v, text, 0);
                error = 2;
            }
            else
            {
                v->server_msg(v, "VNC password ok", 0);
            }
```